rosjava/src/main/java/org/ros/internal/node/service/DefaultServiceClient.java

```diff
@@ -136,15 +134,26 @@ public void connect(final URI uri) {
     Preconditions.checkNotNull(uri, "URI must be specified.");
     Preconditions.checkArgument(uri.getScheme().equals("rosrpc"), "Invalid service URI.");
     Preconditions.checkState(tcpClient == null, "Already connected.");
-    final InetSocketAddress address = new InetSocketAddress(uri.getHost(), uri.getPort());
-    handshakeLatch.reset();
-    tcpClient = tcpClientManager.connect(toString(), address);
+    synchronized (this) {
+      Preconditions.checkState(!triedConnect, "Already called connect.");
+      triedConnect = true;
+    }
     try {
-      if (!handshakeLatch.await(1, TimeUnit.SECONDS)) {
-        throw new RosRuntimeException(handshakeLatch.getErrorMessage());
+      HandshakeLatch handshakeLatch = new HandshakeLatch();
+      serviceClientHandshakeHandler.addListener(handshakeLatch);
+
+      final InetSocketAddress address = new InetSocketAddress(uri.getHost(), uri.getPort());
+      handshakeLatch.reset();
+      tcpClient = tcpClientManager.connect(toString(), address);
+      try {
+        if (!handshakeLatch.await(1, TimeUnit.SECONDS)) {
+          throw new RosRuntimeException(handshakeLatch.getErrorMessage());
+        }
+      } catch (final InterruptedException e) {
+        throw new RosRuntimeException("Handshake timed out.");
       }
-    } catch (final InterruptedException e) {
-      throw new RosRuntimeException("Handshake timed out.");
+    } finally {
+      serviceClientHandshakeHandler.shutdown();
```

Comment:
Should shutdown even if no exception?

Reply:
Yes. The service handshake is only for fetching service info like host ip and port.